sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/BeginInvokePairedWithEndInvoke.cs

```diff
@@ -0,0 +1,230 @@
+            SyntaxKind.DestructorDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.ParenthesizedLambdaExpression,
+            SyntaxKind.SimpleLambdaExpression,
+            SyntaxKind.StructDeclaration
+        }.ToImmutableHashSet();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    var semantic = c.SemanticModel;
+                    var methodSymbol = GetMethodSymbol(invocation, semantic);
+                    if (methodSymbol?.Name == "BeginInvoke" && IsDelegate(methodSymbol))
+                    {
+                        var callbackArg = GetCallbackArg(invocation);
+                        if (callbackArg != null &&
+                            (IsNullLiteral(callbackArg) || !IsCallbackMayContainEndInvoke(callbackArg, semantic)) &&
+                            !IsParentMethodContainsEndInvoke(invocation, semantic))
+                        {
+                            var location = ((SyntaxToken)invocation.GetMethodCallIdentifier()).GetLocation();
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, location));
+                        }
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static bool IsParentMethodContainsEndInvoke(SyntaxNode node, SemanticModel semantic)
+        {
+            var parentContext = node.AncestorsAndSelf()
+                .FirstOrDefault(ancestor => ParentDeclarationKinds.Contains(ancestor.Kind()));
+            return GetEndInvokeList(parentContext, semantic).Count > 0;
+        }
+
+        private ExpressionSyntax GetCallbackArg(InvocationExpressionSyntax invocationExpression)
+        {
+            var callbackArgPos = invocationExpression.ArgumentList.Arguments.Count - 2;
+            var callbackArg = GetArgumentExpressionByNameOrPosition(invocationExpression, "callback", callbackArgPos)
+                ?.RemoveParentheses();
+            return callbackArg;
+        }
+
+        private static bool IsCallbackMayContainEndInvoke(SyntaxNode callbackArg, SemanticModel semantic)
+        {
+            var calledMethod = callbackArg.RemoveParentheses();
+            calledMethod = ReplaceIdentifierByItsInitializer(calledMethod, semantic);
+            calledMethod = ReplaceNewDelegateByMethodDeclaration(calledMethod, semantic);
+            if (calledMethod != null && ParentDeclarationKinds.Contains(calledMethod.Kind()))
+            {
+                return GetEndInvokeList(calledMethod, semantic).Count > 0;
+            }
+
+            return true;
+        }
+
+        private static SyntaxNode ReplaceNewDelegateByMethodDeclaration(SyntaxNode node, SemanticModel semantic)
```

Comment:
The name is confusing - I expect a side effect from the name of the method. Also, please add a comment - which method declaration is this method looking for?

Also , would be clearer if instead of SyntaxNode, it would receive an ObjectCreationExpressionSyntax parameter